### ISSUE TITLE
Add basic functionality for doing deep reinforcement learning

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#1178]
Use case: assume someone has calculated an error externally, perhaps for reinforcement learning.

At present, the workflow is as follows, for each example/minibatch:
- Do forward pass through the network
- Take network outputs, and calculate the error (externally to network)
- Calculate the gradient using MultiLayerNetwork.backpropGradient(INDArray) https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-core/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L2139
- Update the gradient: i.e., apply momentum, l2 regularization, learning rate etc. This is done via something like MultiLayerNetwork.getUpdater().update(MultiLayerNetwork,Gradient,int iteration, int minibatchSize)
  - iteration argument: this value shouldn't matter, unless you are using LR/momentum schedules. In general, it's the number of times you've done parameter updates.
  - minibatchSize: i.e., number of examples in the training minibatch. This is usually the number of rows of your input data.
- Flatten the updated gradient (using Gradient.gradient())
- Get the network parameters
- Do network parameters -= flattened gradient
- Set the network parameters to the new values

This is of course assuming that the user wants to emulate SGD. Using line search methods (conjugate gradient, LBFGS etc) isn't possible via this approach.

At present, the user has to do all of this manually. Ideally, this sort of functionality should be built-in to MultiLayerNetwork.

---------------
Edit: so a few possible designs for doing this properly:
- Provide an 'updateNetworkGivenErrors(INDArray externalError)' method (or similar)
  - This still requires some external code to iterate over examples and calculate errors
  - Can't really use DataSetIterators or anything
- Add a ReinforcementOutputLayer class
  - User would implement an 'ErrorCalculator' interface (or similar) and pass that to the network/config (somehow) to hook into their RL code
  - Upside of this approach: much cleaner design-wise
  - Complications: JSON configuration issues, Spark execution etc.

**Comment by mschaars** [comment]
Clarification question with regards to the call of MultiLayerNetwork.getUpdater().update(MultiLayerNetwork,Gradient,int interation, int minibatchSize):
- What does the iteration parameter here do/what should it be set to? My understanding would be 1, since I compute the gradients sample by sample
- minibatchSize should also be 1 for the same reason?

**Comment by AlexDBlack** [comment]
@mschaars I've added some points to the original post clarifying this.

**Comment by mschaars** [comment]
Thanks! Just one more clarifying question then: 

So the way I implemented this right now is that I go through every sample, calculate the temporal difference and create an error-mask (that fits the output layer) and then call backpropGradient like you described.

I  call the updater function for this error, which in my understanding handles the momentum. Finally, for each sample, I retrieve the updated gradients. I sum up the gradients over each sample in this loop and apply them to the parameters.

Now with regard to the getUpdater.update() call: is my batchsize not 1 in this particular way of doing it, since I call it once per sample? Or, more generally, what exactly does the update call do?

**Comment by AlexDBlack** [comment]
The updater handles momentum (or rmsprop, or adagrad, etc), plus applies l1/l2 regularization, and does division by the minibatch size. In minibatch training (which is what you want to do for computational efficiency, unless you want to update parameters after each example individually), you generally take the average over the example (this is done by providing the batch size to the update method).

**Comment by mschaars** [comment]
Does this code reflect your instructions?:

`   public double update(MiniBatch batch) {

        // Compute Q target for each tuple in the batch
        double[] target = computeTarget(batch);
        double[] difference = new double[target.length];

        for (int i = 0; i < batch.states.length; i++) {
            int actionTaken = batch.actions[i];

            // Get Q value of the action taken in that state-action-reward-nextstate tuple
            double actionValue = evaluate(batch.states[i])[actionTaken];
            difference[i] = target[i] - actionValue;
             difference[i] *= difference[i]

            // Only update the neuron that reflects the action we have taken
            double[] outputMask = new double[actions];
            outputMask[actionTaken] = difference[i];

            Pair<Gradient, INDArray> backpropGradient = trainingNN.backpropGradient(Nd4j.create(outputMask)) ;
            trainingNN.getUpdater().update(trainingNN, backpropGradient.getFirst(), 1, 1);
            Gradient gradient = trainingNN.gradient();

            // Sum up gradients for each sample
            INDArray params = trainingNN.params().dup();

            // Subtract gradients from parameters.
             params.sub(gradient.gradient());

            // Update training network.
            trainingNN.setParams(params);
        }
    }`
And further, could we do the param update only once by adding up the gradient.gradient() values in the loop and then applying them after the loop?

**Comment by AlexDBlack** [comment]
Yes, that looks about right to me. And yes, you could add up the gradient values and do a single parameter update. As it stands though, you are doing parameter updates once per example.

However, if you are going to do batched gradient updates, you might as well do your gradient calculations (using .backpropGradient()) on a matrix (i.e., stack of vectors) instead of a single error vector: this will be more efficient (you'll get parallelization befits from the BLAS code), and it will add the errors automatically for you.

**Comment by mschaars** [comment]
So just for the sake of completion/if anyone wants to use that for an example: 
```
   public double update(MiniBatch batch) {
        int batchSize = batch.actions.length;
        // Get temporal difference targets y= r + gamma * max_a Q(s',a')
        double[] target = computeTarget(batch);
        double[] difference = new double[batchSize];

        // Matrix of error masks
        INDArray errorBatch = Nd4j.create(batchSize, actions);

        for (int i = 0; i < batchSize; i++) {
            int actionTaken = batch.actions[i];

            // Get Q value of the action taken in the tuple
            double actionValue = evaluate(batch.states[i])[actionTaken];

            // Squared temporal difference error for the action taken
            difference[i] = target[i] - actionValue;
            difference[i] *= difference[i];

            // Output mask that contains the error for the action and zeros for all other actions
            double[] outputMask = new double[actions];
            outputMask[actionTaken] = difference[i];

            errorBatch.putRow(i, Nd4j.create(outputMask));
        }

        Pair<Gradient, INDArray> backpropGradient = trainingNN.backpropGradient(errorBatch);

        trainingNN.getUpdater().update(trainingNN, backpropGradient.getFirst(), 1, batchSize);
        Gradient gradient = trainingNN.gradient();
        INDArray params = trainingNN.params().dup();

        params.sub(gradient.gradient());
        trainingNN.setParams(params);

        return errorBatch.meanNumber().doubleValue();
    }
```

**Comment by Joshuaalbert** [comment]
If you wish to take into account the previous state wouldn't you also have to use rnn get/set the previous states before you backprop the error?

**Comment by agibsonccc** [comment]
http://github.com/deeplearning4j/rl4j ?

**Comment by Joshuaalbert** [comment]
I've looked at it, very cool! but I want to implement the A3C different, partly because the software I'm interfacing it to has a very particular layout, and thus wish to get more control over the training (also I noticed rl4j's not quite computing the correct reinforce increments as per Graves 2016 or Williams 1992).

**Comment by agibsonccc** [comment]
Could you file an issue here? It would be great to build something here
rather than having something 1 off. We maintain it. Your code gets used. Rl
gets better. Esp since you are building it anyways.

**Comment by Joshuaalbert** [comment]
Agreed, I will do an issue for RL4J and also give my code when it's working. But it's currently stuck at https://github.com/deeplearning4j/deeplearning4j/issues/2133
